contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+            const leafNodesData = leafNodes.nodes.nodes.filter(
+                (x: AnyNode) => x.tag === 2,
+            );
+            const _side: {
+                price: number;
+                size: number;
+            }[] = leafNodesData
+                .map((x: any) => {
+                    const leafNode: LeafNode = openBookProgram.coder.types.decode(
+                        'LeafNode',
+                        Buffer.from([0, ...x.data]),
+                    );
+                    const size = leafNode.quantity.toNumber();
+                    const price = leafNode.key.shrn(64).toNumber() / 10_000;
+                    return {
+                        price,
+                        size,
+                    };
+                });
+
+            let sortedSide;
+
+            if (side === 'asks') {
+                // Ask side sort
+                sortedSide = _side.sort((
+                    a: { price: number, size: number; },
+                    b: { price: number, size: number; }) => a.price - b.price);
+            } else {
+                // Bid side sort
+                sortedSide = _side.sort((
+                    a: { price: number, size: number; },
+                    b: { price: number, size: number; }) => b.price - a.price);
+            }
+
+            // Aggregate the price levels into sum(size)
+            const _aggreateSide = new Map();
+            sortedSide.forEach((order: { price: number, size: number; }) => {
+                if (_aggreateSide.get(order.price) === undefined) {
+                    _aggreateSide.set(order.price, order.size);
+                } else {
+                    _aggreateSide.set(order.price, _aggreateSide.get(order.price) + order.size);
+                }
+            });
+            // Construct array for our orderbook system
+            let __side: any[][];
+            if (_aggreateSide) {
+                __side = Array.from(_aggreateSide.entries()).map((_side_) => [
+                    (_side_[0].toFixed(4)),
+                    _side_[1],
+                ]);
+            } else {
+                // Return default values of 0
+                return [[0, 0]];
+            }
+            // Update our values for the orderbook
+            if (side === 'asks') {
+                setAsks(__side);
+            } else {
+                setBids(__side);
+            }
+            setLastSlotUpdated(ctx.slot);
+            // Check that we have books
+
+            let tobAsk: number;
+            let tobBid: number;
+
+            // Get top of books
+            if (side === 'asks') {
+                tobAsk = Number(__side[0][0]);
+                // @ts-ignore
+                tobBid = Number(bids[0][0]);
+            } else {
+                // @ts-ignore
+                tobAsk = Number(asks[0][0]);
+                tobBid = Number(__side[0][0]);
+            }
+            // Calculate spread
+            const spread: number = tobAsk - tobBid;
```

Comment:
I think we should put Math.abs() around this as it's periodically displaying wonky numbers... or maybe we need to rethink how this is done. I'm not sure it's catching the update on both books and we really should be doing that, it's my bad for stringing something together which is half working.

Reply:
Added!